src/native/http2_stream_manager.c

```diff
@@ -0,0 +1,403 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "crt.h"
+#include "http_connection_manager.h"
+#include "http_request_response.h"
+#include "http_request_utils.h"
+#include "java_class_ids.h"
+
+#include <jni.h>
+#include <string.h>
+
+#include <aws/common/condition_variable.h>
+#include <aws/common/string.h>
+
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/event_loop.h>
+#include <aws/io/logging.h>
+#include <aws/io/socket.h>
+#include <aws/io/tls_channel_handler.h>
+
+#include <aws/http/connection.h>
+#include <aws/http/http.h>
+#include <aws/http/http2_stream_manager.h>
+#include <aws/http/proxy.h>
+
+/* on 32-bit platforms, casting pointers to longs throws a warning we don't need */
+#if UINTPTR_MAX == 0xffffffff
+#    if defined(_MSC_VER)
+#        pragma warning(push)
+#        pragma warning(disable : 4305) /* 'type cast': truncation from 'jlong' to 'jni_tls_ctx_options *' */
+#    else
+#        pragma GCC diagnostic push
+#        pragma GCC diagnostic ignored "-Wpointer-to-int-cast"
+#        pragma GCC diagnostic ignored "-Wint-to-pointer-cast"
+#    endif
+#endif
+
+/*
+ * Stream manager binding, persists across the lifetime of the native object.
+ */
+struct aws_http2_stream_manager_binding {
+    JavaVM *jvm;
+    jweak java_http2_stream_manager;
+    struct aws_http2_stream_manager *stream_manager;
+};
+
+static void s_destroy_manager_binding(struct aws_http2_stream_manager_binding *binding) {
+    if (binding == NULL) {
+        return;
+    }
+
+    JNIEnv *env = aws_jni_get_thread_env(binding->jvm);
+    if (binding->java_http2_stream_manager != NULL) {
+        (*env)->DeleteWeakGlobalRef(env, binding->java_http2_stream_manager);
+    }
+
+    aws_mem_release(aws_jni_get_allocator(), binding);
+}
+
+static void s_on_http_conn_manager_shutdown_complete_callback(void *user_data) {
+
+    struct aws_http2_stream_manager_binding *binding = (struct aws_http2_stream_manager_binding *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(binding->jvm);
```

Comment:
again, use aws_jni_acquire_thread_env() instead, and check the results

ohhh. maybe you haven't merged from main in a while